File: composeApp/src/commonMain/kotlin/dev/juanrincon/simmerly/welcome/presentation/WelcomeCollage.kt
package dev.juanrincon.simmerly.welcome.presentation

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.sizeIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.ColorFilter
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.layout.layout
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import coil3.compose.AsyncImage
import org.jetbrains.compose.resources.painterResource
import simmerly.composeapp.generated.resources.Res
import simmerly.composeapp.generated.resources.simmerly_logo

private const val COLUMNS = 5
private val GAP = 8.dp

@Composable
internal fun WelcomeCollage(modifier: Modifier = Modifier) {
    BoxWithConstraints(modifier = modifier) {
        val overflow = maxWidth / 8
        val rowWidth = maxWidth + overflow * 2
        val cellSize = (rowWidth - GAP * (COLUMNS - 1)) / COLUMNS

        Row(
            modifier = Modifier.layout { measurable, constraints ->
                val overflowPx = overflow.roundToPx()
                val placeable = measurable.measure(
                    constraints.copy(
                        minWidth = constraints.maxWidth + overflowPx * 2,
                        maxWidth = constraints.maxWidth + overflowPx * 2
                    )
                )
                layout(constraints.maxWidth, placeable.height) {
                    placeable.place(-overflowPx, 0)
                }
            },
            horizontalArrangement = Arrangement.spacedBy(GAP)
        ) {
            Column(verticalArrangement = Arrangement.spacedBy(GAP)) {
                RecipeSquareImage("drawable/recipe_1.jpg", cellSize)
                RecipeSquareImage("drawable/recipe_2.jpg", cellSize)
                ColorBlock(MaterialTheme.colorScheme.tertiary, cellSize)
            }
            Column(verticalArrangement = Arrangement.spacedBy(GAP)) {
                ColorBlock(MaterialTheme.colorScheme.secondaryContainer, cellSize)
                LogoCell(cellSize)
                RecipeCircleImage("drawable/recipe_3.jpg", cellSize)
            }
            Column(verticalArrangement = Arrangement.spacedBy(GAP)) {
                RecipeSquareImage("drawable/recipe_5.jpg", cellSize)
                ColorBlock(MaterialTheme.colorScheme.primary, cellSize)
                RecipeSquareImage("drawable/recipe_6.jpg", cellSize)
            }
            Column(verticalArrangement = Arrangement.spacedBy(GAP)) {
                RecipeSquareImage("drawable/recipe_7.jpg", cellSize)
                RecipeCircleImage("drawable/recipe_8.jpg", cellSize)
                RecipeSquareImage("drawable/recipe_9.jpg", cellSize)
            }
            Column(verticalArrangement = Arrangement.spacedBy(GAP)) {
                RecipeSquareImage("drawable/recipe_10.jpg", cellSize)
                RecipeSquareImage("drawable/recipe_11.jpg", cellSize)
                ColorBlock(MaterialTheme.colorScheme.tertiaryContainer, cellSize)
            }
        }
    }
}

@Composable
private fun RecipeSquareImage(path: String, size: Dp) {
    AsyncImage(
        model = Res.getUri(path), // Need to use Res.getUri(path) see: https://github.com/coil-kt/coil/issues/2812
        contentDescription = null,
        contentScale = ContentScale.Crop,
        modifier = Modifier.size(size).clip(MaterialTheme.shapes.medium)
    )
}

@Composable
private fun RecipeCircleImage(path: String, size: Dp) {
    AsyncImage(
        model = Res.getUri(path), // Need to use Res.getUri(path) see: https://github.com/coil-kt/coil/issues/2812
        contentDescription = null,
        contentScale = ContentScale.Crop,
        modifier = Modifier.size(size).clip(CircleShape)
    )
}

@Composable
private fun ColorBlock(color: Color, size: Dp) {
    Box(
        modifier = Modifier.size(size).background(color, MaterialTheme.shapes.medium)
    )
}

@Composable
private fun LogoCell(size: Dp) {
    Box(
        modifier = Modifier.size(size),
        contentAlignment = Alignment.Center
    ) {
        Image(
            painter = painterResource(Res.drawable.simmerly_logo),
            contentDescription = "Simmerly",
            colorFilter = ColorFilter.tint(MaterialTheme.colorScheme.primary),
            modifier = Modifier.sizeIn(maxWidth = size * 0.6f, maxHeight = size * 0.6f)
        )
    }
}

File: iosApp/iosApp/Features/Welcome/WelcomeCollage.swift
import SwiftUI

/// Decorative image collage ported from welcome/presentation/WelcomeCollage.kt (compact layout only).
struct WelcomeCollage: View {
    private let columns = 5
    private let gap: CGFloat = 8

    var body: some View {
        GeometryReader { geometry in
            let cellSize = (geometry.size.width - gap * CGFloat(columns - 1)) / CGFloat(columns)
            HStack(spacing: gap) {
                VStack(spacing: gap) {
                    RecipeSquareImage("recipe_1", size: cellSize)
                    RecipeSquareImage("recipe_2", size: cellSize)
                    ColorBlock(SimmerlyColor.tertiary, size: cellSize)
                }
                VStack(spacing: gap) {
                    ColorBlock(SimmerlyColor.secondaryContainer, size: cellSize)
                    LogoCell(size: cellSize)
                    RecipeCircleImage("recipe_3", size: cellSize)
                }
                VStack(spacing: gap) {
                    RecipeSquareImage("recipe_5", size: cellSize)
                    ColorBlock(SimmerlyColor.primary, size: cellSize)
                    RecipeSquareImage("recipe_6", size: cellSize)
                }
                VStack(spacing: gap) {
                    RecipeSquareImage("recipe_7", size: cellSize)
                    RecipeCircleImage("recipe_8", size: cellSize)
                    RecipeSquareImage("recipe_9", size: cellSize)
                }
                VStack(spacing: gap) {
                    RecipeSquareImage("recipe_10", size: cellSize)
                    RecipeSquareImage("recipe_11", size: cellSize)
                    ColorBlock(SimmerlyColor.tertiaryContainer, size: cellSize)
                }
            }
        }
    }
}

private struct RecipeSquareImage: View {
    let name: String
    let size: CGFloat

    init(_ name: String, size: CGFloat) {
        self.name = name
        self.size = size
    }

    var body: some View {
        Image(name)
            .resizable()
            .aspectRatio(contentMode: .fill)
            .frame(width: size, height: size)
            .clipShape(RoundedRectangle(cornerRadius: 12))
    }
}

private struct RecipeCircleImage: View {
    let name: String
    let size: CGFloat

    init(_ name: String, size: CGFloat) {
        self.name = name
        self.size = size
    }

    var body: some View {
        Image(name)
            .resizable()
            .aspectRatio(contentMode: .fill)
            .frame(width: size, height: size)
            .clipShape(Circle())
    }
}

private struct ColorBlock: View {
    let color: Color
    let size: CGFloat

    init(_ color: Color, size: CGFloat) {
        self.color = color
        self.size = size
    }

    var body: some View {
        RoundedRectangle(cornerRadius: 12)
            .fill(color)
            .frame(width: size, height: size)
    }
}

private struct LogoCell: View {
    let size: CGFloat

    var body: some View {
        Image("LaunchLogo")
            .renderingMode(.template)
            .resizable()
            .aspectRatio(contentMode: .fit)
            .frame(width: size * 0.6, height: size * 0.6)
            .foregroundStyle(SimmerlyColor.primary)
            .frame(width: size, height: size)
            .accessibilityLabel("Simmerly")
    }
}

#Preview {
    WelcomeCollage()
        .frame(height: 200)
}
